### PR TITLE
Support building against integer-simple

### LINKED ISSUE #135: Compatibility with integer-simple
Currently `store` [hard-depends on `integer-gmp`](https://github.com/fpco/store/blob/247f2863bed0f5c1003ab835075b6dcafb9aa773/package.yaml#L56). Its `.Internal` module [is used](https://github.com/fpco/store/blob/ab518430da6c0b0f79e1e270fb72fa2f6bd7a7d6/src/Data/Store/Internal.hs#L121).

`integer-gmp` implies `gmp` which is LGPL which means currently `store` cannot be used to statically link proprietary products.

It would be great if `store` could offer a flag to use `integer-simple` instead, like `text` (via its `integer-simple` flag) and `cryptonite` (via its `integer-gmp` flag) do.

Could somebody estimate how much effort it would be to write the code that currently uses `import qualified GHC.Integer.GMP.Internals as I` for `integer-simple`?

### PR DESCRIPTION
**Pull request by cocreature**
fixes #135

### CODE REVIEW

File: test/Data/StoreSpec.hs [hunk 1]
```diff
@@ -31,7 +31,9 @@ import           Data.IntSet (IntSet)
 import qualified Data.List.NonEmpty as NE
 import           Data.Map (Map)
 import           Data.Monoid
+#if !MIN_VERSION_primitive(0,7,0)
```
Comment (cocreature) [inline test/Data/StoreSpec.hs]: This is an unrelated change to fix the build against newer versions of `primitive`. It is still not sufficient to get the test suite working on recent GHCs and recent dependencies (I tried 8.6.5) so to test the instance, I’ve simply uncommented everything but `Integer` in the “Store on all monomorphic instances” tests.

**Comment by mgsloan** [comment]
LGTM, thanks!!

**Comment by mgsloan** [comment]
Note: I have some unpushed changes, fixing the other issues, that I will revisit later this evening

**Comment by mgsloan** [comment]
@cocreature I've fixed building of the tests and added explicit unit tests for the serialization of `Integer`: https://github.com/fpco/store/commit/560819846c6920f8e8c47095daa51ce25349bb2b

I'd really appreciate it if you ran the tests with integer-simple.  For some reason I can't seem to build the tests with the flag on, it seems to trigger a variety of different build tool issues:

* With stack-1.9.3, planning fails: https://gist.github.com/mgsloan/ec6d73fa2e28a8488b638ab9225d454c

* With stack-2.2.0, ghc complains about a mismatched hi file, where it seems to expect the version to be included in the hi file: https://gist.github.com/mgsloan/c1ab1688a48b3c9c288153610dc41397

* With cabal new-build: https://gist.github.com/mgsloan/3ddb58af3f4a5405d2ce1e926436a1ff

¯\_(ツ)_/¯ 

Definitely have the flag and extra dep in the stack.yaml https://gist.github.com/mgsloan/d92ed7f28d19ebeb469fdb398a59efc0

**Comment by cocreature** [comment]
@mgsloan Note that you need a different GHC bindist for integer-simple, I got mine via nix. I’ve run the tests (after constraining `primititive` to 0.6.4.0 since the newly added instances break the TH code in the tests). Everything works except for the tests that you have added after my PR. That is expected, I didn’t attempt to match the serialization format used for `integer-gmp` and instead opted for one that naturally fits with `integer-simple`.

In principle it would be possible to match the serialization format of `integer-gmp`, however if that is a goal I think it would be preferable to change the seriailzation format to something that is less dependent on `integer-gmp` internals. Currently the serialization format is very specific to how `integer-gmp` works so given that `integer-simple` works differently (it doesn’t have the small-size optimization, it has a separate 0 value, the digits are an array of words rather than an array of bytes, …) targeting the `integer-gmp` format is quite tricky and probably a fair bit slower than what we have now.

**Comment by mgsloan** [comment]
@cocreature Oh I see.  I thought otherwise because this comment got copied when you wrote the integer-simple code:

```
￼    -- May as well put in the extra effort to use the same encoding as
￼    -- used for the newer integer-gmp.
```

I've pushed changes documenting this mismatch and removing the test

**Comment by mgsloan** [comment]
This change is included in 0.6.1, thanks!
